pkg/requester/types.go

```diff
@@ -47,6 +47,23 @@ type NodeRanker interface {
 	RankNodes(ctx context.Context, job model.Job, nodes []model.NodeInfo) ([]NodeRank, error)
 }
 
+// NodeSelector chooses appropriate nodes for to execute a job.
+type NodeSelector interface {
+	// SelectNodes returns the nodes that should be used to execute the passed job.
+	SelectNodes(context.Context, *model.Job) ([]model.NodeInfo, error)
+	// SelectNodesForRetry returns the nodes that should be used to retry the
+	// passed failed executions in the context of the passed job. If no nodes
+	// are returned, the executions do not need to be retried just yet.
+	SelectNodesForRetry(context.Context, *model.Job, *model.JobState) ([]model.NodeInfo, error)
+	// SelectBids returns the pending bids on the passed job that should be
+	// accepted or rejected.
+	SelectBids(context.Context, *model.Job, *model.JobState) (accept, reject []model.ExecutionState)
```

Comment:
nit: by virtue of the interface name it is implied these methods select some set of nodes, these methods could be shortened to:
`Nodes`
`NodesForRetry`
`Bids`

Reply:
Agreed... I think the `Selector` part of the name is the part that I am not happy with. It now seems to be doing more than just "selecting". It will:

* Decide whether there are enough nodes to run the job, and fail it if not
* Select the nodes that should run the job
* Decide whether we are ready to execute the job (i.e. did we get enough bids)
* Select the bids that should be accepted
* Decide whether the job is complete enough to report success
* Select which executions should be retried, and the nodes to retry them on

So selection is a big part of it, but now it is also encoding the "logic of the deal", i.e. ensuring that the "deal" that the user requested is fulfilled or failing if not. I almost want to call it a very whimsical `Dealmaker` but I don't think that is an improvement in clarity!